File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseHelper(self, s: ListNode, k: int):
        l = s.next
        r = l
        for _ in range(k):
            if r:
                r = r.next
            else:
                return
        
        curr, prev = l, r
        while curr != r:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        s.next = prev

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        s = dummy
        while True:
            self.reverseHelper(s, k)
            for _ in range(k):
                if s.next:
                    s = s.next
                else:
                    return dummy.next

        

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */

function reverseHelper(s, k) {
    let l = s.next
    let r = l;
    for (let i = 0; i < k; i++) {
        if (r) {
        r = r.next;
        } else {
            return
        }
    }
    let curr = l, prev = r
    while(curr !== r) {
        let next = curr.next
        curr.next = prev
        prev = curr
        curr = next
    }
    s.next = prev
}
var reverseKGroup = function(head, k) {
    let dummy = new ListNode();
    dummy.next = head
    let s = dummy;
    while(true) {
                console.log(s)
        reverseHelper(s, k)
                console.log(s)
        for (let i = 0; i < k; i++) {
            if (s.next) {
                s = s.next
            } else {
                console.log(s)
                return dummy.next
            }
        }
    }
};